Fix unhandled exception when UWP ClientWebSocket loses connection to server

Diego-Perez-Botero opened the pull request:
The UWP implementation of ClientWebSocket uses WinRT MessageWebSocket APIs internally. If a connection-related or message-processing failure occurs, WinRT raises a MessageReceived event that throws an exception containing the failure's details when GetDataReader() is called. Unfortunately, ClientWebSocket isn't handling those exceptions inside its MessageReceived event handler, so they end up bringing down the entire process.

These changes add basic handling of GetDataReader() exceptions. They also contain a new functional test that pins the current behavior of all ClientWebSocket implementations around the "connection is prematurely closed" scenario. Without the proposed product code changes, this new test case reproduces the crash tracked by #17317.

Fixes #17317

Diego-Perez-Botero commented:
@dotnet-bot test outerloop netcoreapp OSX10.12 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Release

Diego-Perez-Botero commented:
@dotnet-bot help

davidsh commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs:
```diff
@@ -279,47 +279,60 @@ public override void Dispose()
+                    }
                 }
             }
+            catch (Exception)
```
We usually do not catch all exceptions in .NET code because there are cases where we want "unexpected" exceptions to crash the process.  So, perhaps there is a more specific type of exception to catch here?

Also, we probably should save this exception in the catch block and then use it as an inner exception when you finally throw the WebSocketException later on. This would greatly help with diagnosing these failures.

Diego-Perez-Botero replied:
Sadly, WinRT WebSockets always throw exceptions of type "Exception". However, I can determine whether or not we're dealing with a known error by using the [WebSocketError.GetStatus](https://docs.microsoft.com/en-us/uwp/api/Windows.Networking.Sockets.WebSocketError) method. Any unknown errors should still crash the process.

The inner exception idea is great. I'll definitely do that :)

davidsh replied:
>WinRT WebSockets always throw exceptions of type "Exception"

I thought WinRT threw COMException.  Maybe that can be used?

Diego-Perez-Botero replied:
I just tried catching COMException, but it didn't work :(

From a debugger, the first chance exception does appear to be of type "System.Exception"; nothing more specific than that:

>  0:007> !pe
> Exception object: 0000026890006000
> Exception type:   System.Exception
> Message:          Exception from HRESULT: 0x80072EFE
> InnerException:   <none>
> StackTrace (generated):
> <none>
> StackTraceString: <none>
> HResult: 80072efe

davidsh replied:
ok.  In the past, it used to be COMException. But that might have changed over time.

Diego-Perez-Botero replied:
I'm now propagating the WinRT exception as an inner exception within a custom WebSocketException to any pending ReceiveAsync/SendAsync operations. Subsequent ReceiveAsync/SendAsync operations fail with the typical "invalid state" WebSocketException. This mirrors the .NET Framework behavior, as can be observed in the updated test case.

davidfowl commented on src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs:
```diff
@@ -279,47 +279,60 @@ public override void Dispose()
+
+                        WebSocketReceiveResult recvResult = new WebSocketReceiveResult((int) readCount, messageType,
+                            endOfMessage);
+                        _webSocketReceiveResultTcs.TrySetResult(recvResult);
```
Is the intent to run the continuation on this thread? These TCS's probably need `TaskCreationOptions.RunContinuationsAsynchronously`

Diego-Perez-Botero replied:
I do believe that we want the continuation to run on this thread, considering that:

- ClientWebSocket supports at most one ReceiveAsync operation at any given time.
- WinRT invokes MessageReceived events in serial fashion, waiting for our handler to finish executing before raising the next event.

If we were to use the `TaskCreationOptions.RunContinuationsAsynchronously` option, we may end up processing MessageReceived events at a higher rate than what the .NET app is able to handle via ReceiveAsync, which would lead to increased memory footprint (internally-buffered data would be ever-increasing in size).

Diego-Perez-Botero commented:
@dotnet-bot test outerloop netcoreapp OSX10.12 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Release

Diego-Perez-Botero commented:
I have addressed all comments so far and the UWP ClientWebSocket exception-related behavior now closely mirrors that of the .NET Framework implementation.

Diego-Perez-Botero commented:
@dotnet-bot test outerloop netcoreapp OSX10.12 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Release

Diego-Perez-Botero commented:
@dotnet-bot test Linux x64 Release Build

Diego-Perez-Botero commented:
The _**OuterLoop netcoreapp Windows_NT Debug x64**_ test failures are unrelated to this PR's changes. They are being tracked by #21554,

stephentoub commented on src/System.Net.WebSockets.Client/tests/SendReceiveTest.cs:
```diff
@@ -349,5 +351,124 @@ public class SendReceiveTest : ClientWebSocketTestBase
+
+                    const uint WININET_E_CONNECTION_RESET = 0x80072eff;
+
+                    Win32Exception pendingReceiveException = await Assert.ThrowsAnyAsync<Win32Exception>(() => pendingReceiveAsync);
```
Why is this WinHttpWebSocket behavior being accepted and codified in a test rather than it being a bug?  All of the other implementations are throwing a WebSocketException, but this is throwing a Win32Exception?
cc: @davidsh

Diego-Perez-Botero replied:
My objective with this test was to pin the behaviors of the various WebSocketClient.ReceiveAsync implementations, though the WinHttpWebSocket behavior codified here is essentially a repro of #14103.

davidsh replied:
Instead of this being a test, it should have been separated out and marked as ActiveIssue against #14103.
